rfcs/rfcTODO.md

```diff
@@ -0,0 +1,142 @@
+
+## Motivation
+
+Language interoperability concerns sometimes lead to situations where it is necessary to represent typed data - such as strings or numbers - in a way that can be reliably distinguished. Elsewhere in Perl there are attempts to add this distinction. This RFC attempts to address how to handle values of a boolean type; values that represent a simple true-or-false nature.
+
+For example, JSON and MsgPack are commonly-encountered serialisation formats that Perl can generate and parse, though in both cases while the formats themselves can represent boolean truth values distinctly from numbers or strings, perl cannot preserve that distinction.
+
+There is a tendancy for modules to make up this shortfall with workarounds like the `JSON::PP::Boolean` type:
+
+```
+$ perl -MJSON::MaybeUTF8=decode_json_utf8 -MData::Dump=pp 
+   -E 'say pp(decode_json_utf8($ARGV[0]))' '[1,true]'
+
+[1, bless(do{\(my $o = 1)}, "JSON::PP::Boolean")]
+```
+
+Obviously such a solution is specific to JSON encoding and does not apply to, for example, message gateway between JSON and MsgPack, which would require some translation inbetween. A true in-core solution to this problem would have many benefits to interoperability of data handling between these various modules.
+
+((TODO: Add some comments about purely in-core handling as well that don't rely on serialisation))
+
+## Rationale
+
+## Specification
+
+There are two parts to this specification; the lower-level in-core details of the C code implementation that are visible to XS code; and the higher-level Perl-visible details that are made visible to Perl programs.
+
+### C-level Internals
+
+At the C level, SVs will require a flag (or some other detail about them) that can reliably convey that they contain a boolean. It may be possible to find a spare `SvFLAGS` field for this purpose, at which point there would be macros on a theme similar to the `SvPOK`/`SvIOK`/etc.. set. I would suggest avoiding the single-letter abbreviations of the older style of code - letters aren't so expensive these days that we can't at least spell out the word "BOOL" in full.
+
+```
+SvBOOLOK(sv)
+SvBOOLOK_on(sv)
+SvBOOLOK_off(sv)
+```
+
+The core immortals `PL_sv_yes` and `PL_sv_no` will always respond true to `SvBOOLOK()`, and such a flag will be reliably copied by `sv_setsv()` and friends. The result here is that the result of boolean-returning ops will be `SvBOOLOK()` and this flag will remain with any copies of that value that get made - either over the arguments stack or stored in lexicals or elements of aggregate structures.
+
+These `SvBOOLOK()` values will still be subject to the usual semantics regarding macros like `SvPV` or `SvIV` - so numerically they will still be 1 or 0, and stringily they will still be "1" and "" (though see the Future Scope section below).
```

Comment:
You might want to clarify here to cover @iabyn's comment, that the POK, IOK etc flags on the SV won't be changing, so code like pp_add doesn't need any changes.